File: packages/composer/src/Contract.php
<?php

declare(strict_types=1);

namespace Escalated\Contract;

use JsonException;
use RuntimeException;

/**
 * Reads the published contract.
 *
 * The point of this class is that a port never keeps its own copy of the
 * manifest. Before this package existed, each of the eleven ports vendored
 * escalated-pages.json into its own test fixtures and refreshed it by hand.
 * Two of them drifted and their parity tests stayed green anyway, because each
 * port was comparing itself against its own stale copy.
 *
 * Read from here instead, and a `composer update` is what surfaces the drift.
 */
final class Contract
{
    /** @var array{version: string, pages: list<string>, props: array<string, array{props: list<string>, required: list<string>}>}|null */
    private static ?array $pages = null;

    /**
     * Every page name the shared frontend resolves.
     *
     * A name a backend renders that is not in here is not an error at runtime.
     * Inertia returns 200, the resolver returns undefined, and the panel comes
     * up blank, which reads as a permissions problem or an empty dataset.
     *
     * @return list<string>
     */
    public static function pages(): array
    {
        return self::manifest()['pages'];
    }

    /**
     * The props a page reads, and which of them it cannot render without.
     *
     * @return array{props: list<string>, required: list<string>}
     */
    public static function propsFor(string $page): array
    {
        $props = self::manifest()['props'];

        if (! isset($props[$page])) {
            throw new RuntimeException(sprintf(
                'No props recorded for page "%s". Either the name is wrong, or the frontend has not released it yet.',
                $page
            ));
        }

        return $props[$page];
    }

    /**
     * True when the shared frontend has a component for this name.
     */
    public static function rendersPage(string $page): bool
    {
        return in_array($page, self::pages(), true);
    }

    /**
     * The @escalated-dev/escalated release this manifest came from.
     */
    public static function frontendVersion(): string
    {
        return self::manifest()['version'];
    }

    /**
     * Absolute path to a conformance fixture directory, for a runner to walk.
     */
    public static function conformancePath(string $area = ''): string
    {
        $path = self::contractDir().'/conformance';

        return $area === '' ? $path : $path.'/'.$area;
    }

    /**
     * @return array{version: string, pages: list<string>, props: array<string, array{props: list<string>, required: list<string>}>}
     */
    private static function manifest(): array
    {
        if (self::$pages !== null) {
            return self::$pages;
        }

        $path = self::contractDir().'/pages.json';
        $raw = @file_get_contents($path);

        if ($raw === false) {
            throw new RuntimeException('Cannot read the page manifest at '.$path);
        }

        try {
            /** @var array{version: string, pages: list<string>, props: array<string, array{props: list<string>, required: list<string>}>} $decoded */
            $decoded = json_decode($raw, true, 512, JSON_THROW_ON_ERROR);
        } catch (JsonException $e) {
            throw new RuntimeException('The page manifest at '.$path.' is not valid JSON: '.$e->getMessage(), 0, $e);
        }

        return self::$pages = $decoded;
    }

    private static function contractDir(): string
    {
        return dirname(__DIR__).'/contract';
    }
}

File: packages/go/contract.go
// Package contract reads the published Escalated contract.
//
// A port should not keep its own copy of the manifest. Each of the eleven
// ports used to vendor escalated-pages.json into its own testdata and refresh
// it by hand; two drifted and their parity tests stayed green, because each
// port was comparing itself against its own stale copy.
//
// The contract files are embedded, so a consumer needs nothing on disk.
package contract

import (
	"embed"
	"encoding/json"
	"fmt"
	"io/fs"
	"sync"
)

//go:embed contract
var files embed.FS

// PageProps is what a single page component reads.
type PageProps struct {
	// Props the component reads.
	Props []string `json:"props"`
	// Props it cannot render without.
	Required []string `json:"required"`
}

// Manifest is the published page contract.
type Manifest struct {
	// Version is the @escalated-dev/escalated release this came from.
	Version string `json:"version"`
	// Pages is every page name the shared frontend resolves.
	Pages []string `json:"pages"`
	// Props is keyed by page name.
	Props map[string]PageProps `json:"props"`
}

var (
	once   sync.Once
	loaded Manifest
	loadIn error
)

func load() (Manifest, error) {
	once.Do(func() {
		raw, err := files.ReadFile("contract/pages.json")
		if err != nil {
			loadIn = fmt.Errorf("reading embedded page manifest: %w", err)
			return
		}
		if err := json.Unmarshal(raw, &loaded); err != nil {
			loadIn = fmt.Errorf("parsing embedded page manifest: %w", err)
		}
	})
	return loaded, loadIn
}

// Pages returns every page name the shared frontend resolves.
//
// A name a backend renders that is not here is not a runtime error. Inertia
// returns 200, the resolver returns undefined, and the panel comes up blank.
func Pages() ([]string, error) {
	m, err := load()
	if err != nil {
		return nil, err
	}
	return m.Pages, nil
}

// PropsFor returns the props a page reads.
func PropsFor(page string) (PageProps, error) {
	m, err := load()
	if err != nil {
		return PageProps{}, err
	}
	p, ok := m.Props[page]
	if !ok {
		return PageProps{}, fmt.Errorf(
			"no props recorded for page %q: either the name is wrong, or the frontend has not released it yet", page)
	}
	return p, nil
}

// RendersPage reports whether the shared frontend has a component for a name.
func RendersPage(page string) (bool, error) {
	m, err := load()
	if err != nil {
		return false, err
	}
	for _, p := range m.Pages {
		if p == page {
			return true, nil
		}
	}
	return false, nil
}

// FrontendVersion returns the frontend release this manifest came from.
func FrontendVersion() (string, error) {
	m, err := load()
	if err != nil {
		return "", err
	}
	return m.Version, nil
}

// Conformance returns the embedded conformance fixtures, for a runner to walk.
func Conformance() (fs.FS, error) {
	return fs.Sub(files, "contract/conformance")
}
